Feature/co 472 integrate zonos elements

nick-lanners opened the pull request:
# CO-472: Integrate Zonos Elements

This pull request implements the integration of Zonos Elements into our commerce platform, providing cross-border commerce functionality for international shoppers.

## Changes

- Replaced Shopify API integration with Zonos Elements API
- Implemented Zonos Hello for accurate duty and tax calculations at the product level
- Implemented client-side Zonos script loading and initialization
- Added server-side API endpoints for cart management
- Created type definitions for Zonos API responses
- Built comprehensive unit tests for cart actions and API functionality
- Added documentation and specifications for the Zonos integration

## Technical Details

- Removed legacy Shopify integration files and types
- Created a new API client for interacting with Zonos Elements API
- Implemented environment-specific API endpoints configuration
- Set up proper authentication using CUSTOMER_GRAPH_TOKEN for server-side and NEXT_PUBLIC_ZONOS_API_KEY for client-side
- Added ZonosLayoutSetup component for initializing Zonos Elements
- Updated cart context to use Zonos cart APIs

## Testing

- Added unit tests for cart actions
- Added tests for Zonos API configuration
- Added tests for Zonos cart operations
- Tested client-side initialization in development environment

goodpickle committed:
Initial commit

goodpickle committed:
Add initial project structure and documentation for Zonos Commerce

- Created README.md with project overview, setup instructions, and development guidelines.
- Added SPECS.md detailing technical specifications and architecture.
- Implemented Zonos API client and configuration files for integration with Zonos Elements.
- Established project structure following Next.js Commerce patterns, including components and specs directories.
- Included deployment specifications and testing strategies for comprehensive project management.

nick-lanners committed:
Merge pull request #1 from Zonos/build/initial-specs

Add initial project structure and documentation for Zonos Commerce

Thienhuynh95 committed:
build: initial commit of Next.js Commerce application, including essential configuration files, environment setup, and foundational components for the e-commerce platform.

Thienhuynh95 committed:
Merge branch 'main' of github.com:Zonos/commerce into build/init-project

Thienhuynh95 committed:
feat: add package json

Thienhuynh95 committed:
refactor: add github action for type checking

Thienhuynh95 committed:
feat: add prettier ignore

Thienhuynh95 committed:
refactor: remove const

Thienhuynh95 committed:
refactor: rerun prettier and re-organize api server fetch

Thienhuynh95 committed:
refactor: re-run prettier

Thienhuynh95 committed:
refactor: use test URL with functionality

Thienhuynh95 committed:
refactor: re-run prettier

Thienhuynh95 committed:
build: add package manager

Thienhuynh95 committed:
ci: update workflow to include Node.js setup and pnpm action

Thienhuynh95 committed:
ci: add pnpm install step to CI workflow

Thienhuynh95 committed:
refactor: remove welcome toast

goodpickle committed:
Merge pull request #2 from Zonos/build/init-project

Build: Project Setup and API Integration Improvements

nick-lanners committed:
feat(api): implement Zonos API integration types and endpoints

Add API integration specification with Zonos Elements API endpoints, authentication details, and comprehensive TypeScript type definitions for cart operations. Update pnpm lockfile with new dependencies including graphql-request, fuse.js, and turndown.

nick-lanners committed:
fix: prettier format

nick-lanners committed:
feat(testing): Implement Vitest for unit testing

This commit introduces a comprehensive testing infrastructure with Vitest for the commerce project:

- Added vitest.config.ts configuration for Node.js environment
- Created test directory structure with README.md explaining testing approach
- Implemented unit tests for core utility functions:
  - type-guards.test.ts: Testing isObject and isShopifyError functions
  - utils.test.ts: Testing createUrl, ensureStartsWith, and validateEnvironmentVariables
  - zonos-api-config.test.ts: Testing API endpoint construction and environment-specific URL selection
  - zonos-cart.test.ts: Testing cart retrieval functionality
  - zonos-fetch.test.ts: Testing API client fetch operations

- Updated testing-strategy.spec.md to reflect the current implementation status
- Removed outdated cart-actions.test.ts that referenced a non-existent file path

These changes establish a solid foundation for testability with proper mocking approaches and complete test coverage for utility functions.

nick-lanners committed:
fix: types and linting

nick-lanners committed:
fix: linting, remove console logs

nick-lanners committed:
test: cart actions tests

nick-lanners committed:
Merge branch 'main' of https://github.com/Zonos/commerce into feature/co-472-integrate-zonos-elements

nick-lanners committed:
fix: remove shopify files

nick-lanners committed:
fix: prettier

nick-lanners committed:
fix: any types

nick-lanners committed:
fix: remove unused type-guards

nick-lanners committed:
fix: change Zonos implementation to use CDN

nick-lanners committed:
fix: prettier and tsc

nick-lanners committed:
fix: optimize script, adjust api routes

nick-lanners committed:
fix: prettier

nick-lanners committed:
fix: adjust ZonosScript to resolve mismatched client and server

nick-lanners committed:
fix: prettier

nick-lanners committed:
fix: clean up project to more closely match the specs

nick-lanners committed:
fix: prettier and lint

goodpickle reviewed (changes requested):
Project Analysis
After reviewing the codebase, I've identified a few issues with the environment variable management:

Issues
1. Duplicate Environment Validation
The project has two separate environment validation systems that overlap:

A simple implementation in utils.ts with [validateEnvironmentVariables()](vscode-file://vscode-app/private/var/folders/_1/ynzjyw9x3mz2bd14v0g9gg3r0000gn/T/AppTranslocation/F5F431DC-BEF1-4A5D-9191-57F6D953807E/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html)
A more robust system using @t3-oss/env-core and Zod in the [environment](vscode-file://vscode-app/private/var/folders/_1/ynzjyw9x3mz2bd14v0g9gg3r0000gn/T/AppTranslocation/F5F431DC-BEF1-4A5D-9191-57F6D953807E/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) directory
The simple validation in [utils.ts](vscode-file://vscode-app/private/var/folders/_1/ynzjyw9x3mz2bd14v0g9gg3r0000gn/T/AppTranslocation/F5F431DC-BEF1-4A5D-9191-57F6D953807E/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) checks for required variables [CUSTOMER_GRAPH_TOKEN](vscode-file://vscode-app/private/var/folders/_1/ynzjyw9x3mz2bd14v0g9gg3r0000gn/T/AppTranslocation/F5F431DC-BEF1-4A5D-9191-57F6D953807E/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) and [SITE_NAME](vscode-file://vscode-app/private/var/folders/_1/ynzjyw9x3mz2bd14v0g9gg3r0000gn/T/AppTranslocation/F5F431DC-BEF1-4A5D-9191-57F6D953807E/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html), but these are also validated by the T3 Env setup. This duplication creates:

Potential inconsistencies between validation rules
Double validation that could lead to confusing error messages
Maintenance burden when adding new environment variables
2. Missing Validation for Required Variables
Based on [.env.example](vscode-file://vscode-app/private/var/folders/_1/ynzjyw9x3mz2bd14v0g9gg3r0000gn/T/AppTranslocation/F5F431DC-BEF1-4A5D-9191-57F6D953807E/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html), [ZONOS_REVALIDATION_SECRET](vscode-file://vscode-app/private/var/folders/_1/ynzjyw9x3mz2bd14v0g9gg3r0000gn/T/AppTranslocation/F5F431DC-BEF1-4A5D-9191-57F6D953807E/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) appears to be required, but the simple validation in [utils.ts](vscode-file://vscode-app/private/var/folders/_1/ynzjyw9x3mz2bd14v0g9gg3r0000gn/T/AppTranslocation/F5F431DC-BEF1-4A5D-9191-57F6D953807E/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) doesn't check for it (though the T3 Env setup does).

3. Inconsistent Error Handling
The two validation approaches handle errors differently:

[utils.ts](vscode-file://vscode-app/private/var/folders/_1/ynzjyw9x3mz2bd14v0g9gg3r0000gn/T/AppTranslocation/F5F431DC-BEF1-4A5D-9191-57F6D953807E/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) throws a simple error with a list of missing variables
The T3 Env setup provides more detailed errors through Zod validation
Recommendations
Remove the redundant validation in [utils.ts](vscode-file://vscode-app/private/var/folders/_1/ynzjyw9x3mz2bd14v0g9gg3r0000gn/T/AppTranslocation/F5F431DC-BEF1-4A5D-9191-57F6D953807E/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html): The T3 Env approach is more robust and type-safe. The [validateEnvironmentVariables()](vscode-file://vscode-app/private/var/folders/_1/ynzjyw9x3mz2bd14v0g9gg3r0000gn/T/AppTranslocation/F5F431DC-BEF1-4A5D-9191-57F6D953807E/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) function should be removed or refactored to use the T3 Env validation.

Consolidate all environment validation in the T3 Env setup: Any custom validation logic from [utils.ts](vscode-file://vscode-app/private/var/folders/_1/ynzjyw9x3mz2bd14v0g9gg3r0000gn/T/AppTranslocation/F5F431DC-BEF1-4A5D-9191-57F6D953807E/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) (like platform validation) should be moved to the appropriate environment schema files.

Ensure consistency with [.env.example](vscode-file://vscode-app/private/var/folders/_1/ynzjyw9x3mz2bd14v0g9gg3r0000gn/T/AppTranslocation/F5F431DC-BEF1-4A5D-9191-57F6D953807E/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html): Review all required variables in [.env.example](vscode-file://vscode-app/private/var/folders/_1/ynzjyw9x3mz2bd14v0g9gg3r0000gn/T/AppTranslocation/F5F431DC-BEF1-4A5D-9191-57F6D953807E/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) and confirm they're properly validated in the environment schemas.

Update tests: The [utils.test.ts](vscode-file://vscode-app/private/var/folders/_1/ynzjyw9x3mz2bd14v0g9gg3r0000gn/T/AppTranslocation/F5F431DC-BEF1-4A5D-9191-57F6D953807E/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) file contains tests for the now-redundant [validateEnvironmentVariables()](vscode-file://vscode-app/private/var/folders/_1/ynzjyw9x3mz2bd14v0g9gg3r0000gn/T/AppTranslocation/F5F431DC-BEF1-4A5D-9191-57F6D953807E/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) function. These tests should be refactored to validate the T3 Env setup instead.

This will create a more maintainable and consistent approach to environment variable validation throughout the project.

goodpickle commented:
tests/README.md has references to `npm` instead of `pnpm`

goodpickle commented:
Let's switch to a main version of Next. The latest (non-canary) version.

Thienhuynh95 commented on .env.example:
```diff
@@ -1,5 +1,21 @@
+
+# Optional client-side environment variables
+NEXT_PUBLIC_ZONOS_ENVIRONMENT="sandbox" # or "production"
+NEXT_PUBLIC_COMPANY_NAME="Your Company Name"
```
I don't see this NEXT_PUBLIC_ZONOS_ENVIRONMENT is being used any where. What's this one for?

nick-lanners replied:
This, and a couple other env vars, are defined in the specs but haven't been implemented in the project yet as we felt they were future features. Should we remove these until we implement them or add comments that they will be implemented in a future release?

Thienhuynh95 replied:
yah adding a comment to it would be great.

Thienhuynh95 commented on specs/project-setup.spec.md:
````diff
@@ -0,0 +1,255 @@
+### .eslintrc.js
+
+ESLint should be configured to use Zonos's standard configuration:
+
+```javascript
+module.exports = {
+  extends: ['zonos'],
+  parserOptions: {
+    project: './tsconfig.json'
+  }
+}
````
I don't see eslint in this project atm, is that the plan getting it in in the future?

nick-lanners replied:
I thought you had purposely set up the project to not use eslint for a reason. Was I mistaken? I can set it up.

Thienhuynh95 replied:
You're right but that's not what I meant. I just mean we have it documented that this project is using eslint, but no eslint footprint anywhere.

Thienhuynh95 replied:
so we either consider to remove it from the docs or plan to add it in in the future.

goodpickle replied:
Let's get eslint in the project. Make sure to use the latest flat config so we don't need to update it later.

nick-lanners replied:
Alrighty. I had Agent set up a basic eslint config. How extensive do we want to be with this? I know our other projects have huge eslint configs, but they mix prettier and eslint stuff together, which we have expressed wanting to get away from?

nick-lanners committed:
fix: cleanup and adjustment

adjust environment variables to remove duplicates
add eslint
fix linting errors across project

nick-lanners committed:
fix: add not-found file, github workflows

nick-lanners committed:
fix: adjust how env var is validated

nick-lanners committed:
fix: prettier
